static/docs/user-guide/using-docker.md

```diff
@@ -0,0 +1,108 @@
+```
+
+For a recommended reading about writing Dockerfiles for Python applications,
+you can take a look at https://pythonspeed.com/docker/
+
+## Using Docker for your development environment
+
+Since `dvc` is a command line application, we have several recommendiations
+to enhance the experience:
+  * Make sure your _locale_ is set to UTF-8
+  * Install the `Zsh` shell and its respective autocompletion script.
+  * Install `less` so you can pipe the output of `dvc pipeline show`.
+  * Enable movements with `Ctrl + Arrow` by editing the _inputrc_ file.
+  * Use `pip install dvc` at the end of the file for better caching.
+  * Set your `WORKDIR` to the path where you are going to mount your project.
+
+If you are using Linux, remember to map your user to your container. This way,
+you make sure your files are still writtable. For more information:
+https://docs.docker.com/engine/security/userns-remap/
+
+### Dockerfile
+
+The following is a suggestion for a Dockerfile to use during development:
+
+```dockerfile
+FROM python
+
+# Use UTF-8 as the default locale
+ENV LANG=C.UTF-8
+
+# Better movement with [Ctrl + →] and [Ctrl + ←]
+ENV INPUTRC=/etc/inputrc
+
+RUN set -ex \
+  && echo '"\e[1;5C": forward-word' >> /etc/inputrc \
+  && echo '"\e[1;5D": backward-word' >> /etc/inputrc
+
+# Set working directory.
+# Mount your project under the same path.
+WORKDIR /usr/src
+
+# Install useful programs
+RUN apt-get update -y && apt-get install -y \
+      zsh \
+      less \
+      git
+
+#  Set Zsh as the default shell
+SHELL ["/usr/bin/zsh", "-c"]
+
+# Enable autocompletion scripts for DVC
+RUN wget \
+      -O /usr/local/share/zsh/site-functions/_dvc \
+      https://raw.githubusercontent.com/iterative/dvc/master/scripts/completion/dvc.zsh \
+      && echo "autoload -U compinit && compinit" >> /root/.zshrc \
+      && echo "zstyle ':completion:*:*:*:*:*' menu select" >> /root/.zshrc \
+      && echo "zstyle ':completion:*:matches' group 'yes'" >> /root/.zshrc \
+      && echo "zstyle ':completion:*:options' description 'yes'" >> /root/.zshrc \
+      && echo "zstyle ':completion:*:options' auto-description '%d'" >> /root/.zshrc \
+      && echo "zstyle ':completion:*:corrections' format ' %F{green}-- %d (errors: %e) --%f'" >> /root/.zshrc \
+      && echo "zstyle ':completion:*:descriptions' format ' %F{yellow}-- %d --%f'" >> /root/.zshrc \
+      && echo "zstyle ':completion:*:messages' format ' %F{purple} -- %d --%f'" >> /root/.zshrc \
+      && echo "zstyle ':completion:*:warnings' format ' %F{red}-- no matches found --%f'" >> /root/.zshrc \
+      && echo "zstyle ':completion:*:default' list-prompt '%S%M matches%s'" >> /root/.zshrc \
+      && echo "zstyle ':completion:*' format ' %F{yellow}-- %d --%f'" >> /root/.zshrc \
+      && echo "zstyle ':completion:*' group-name ''" >> /root/.zshrc \
+      && echo "zstyle ':completion:*' verbose yes" >> /root/.zshrc
+
+# Install DVC with s3 enabled
+RUN pip install "dvc[s3]"
```

Comment:
why s3 only?

probably not the best idea to install into the global space - we advertise a pretty bad practice with this

why don't we use conda, like @casperdcl suggested in one of the tickets?

Reply:
Specifically `$CONDA_PREFIX/etc/conda/activate.d/` and `deactivate.d/`

Reply:
> probably not the best idea to install into the global space - we advertise a pretty bad practice with this

@shcheklein , why do you consider it a bad idea?
I don't see any problem at all, to be honest; a docker container should only be dealing with one stuff, anyways.

> why s3 only?
@shcheklein , I was about to use `dvc[all]` but `pyarrow` requires `CPython` and it is not supporten on Python 3.8 yet; I decided to leave it as an exercise to the reader, the idea is not to provide _the best_ Dockerfile, but just an example to display the bigger picture.

> why don't we use conda, like @casperdcl suggested in one of the tickets?

There's a note already about installing DVC through conda, an example with two Dvcfiles  seems excessive. I'm trying to cover just the basics.

Reply:
I would think *the* best `Dockerfile` is exactly what we need to advocate. Otherwise we're saying "oh yeah we know of the existence of this thing you call docker and you can probably use it with our stuff; we've made a page about it on our website but couldn't be bothered to make it fully correct. Here haz some vaguely not-quite-correct musing of how it might sort of look. k's thx"

Reply:
python 3.7 instead of 3.8? Don't see any need to use bleeding edge...

Reply:
@casperdcl , I see your point; but I don't think there's a **correct** image :slightly_frowning_face: . It may vary from user to user, this is sort of like a baseline.
Also, this Dockerfile is for _development only_, "production" is such a hard thing to get right.

Reply:
@casperdcl , don't get me wrong, I'd be happy to have a _more sound_ version of this image)

open to discussion and commits, as always)!

Reply:
@mroutis so why should it be Python 3.8 then, let's install 3.7.

> why do you consider it a bad idea?

Probably not a big deal for Docker - since it's already an isolated environment. It's just publishing it this way in docs can be considered as an instruction to do the same outside Docker, so was thinking if we should use apt-get for example, or `conda` or some other sane way to deliver the tool. May be it's just my bias - "never ever do sudo pip install something" :)

Reply:
@shcheklein , we are running with `root` :sweat_smile: , we would need to start by creating a specific user for running the application, installing `sudo` and then add the user to the `sudo` group. Also, allowing to sudo without password, so editing the sudoers list :grimacing: 

Not sure if this is the best practice, would need to research

Reply:
I think Ivan meant to NOT ever use `sudo` before `pip` (on global env). And suggested installing DVC under a `virtualenv`, even if its a Docker image. I don't think that would be useful though, because you kind of expect your entire container to have DVC available everywhere if you're installing it that way.